gremlin-go/design.md

```diff
@@ -0,0 +1,325 @@
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+
+-->
+
+# Gremlin-Go Design
+
+## Entities
+
+Gremlin-Go, named `gremlingo` as a package, consists of two main groups of entities at a high level: *Driver*-related and *Gremlin*-related entities. 
+
+### Driver
+
+Driver-related entities are used to handle the processing, parsing, and sending of Gremlin traversal queries. They are also responsible for deserializing responses from a Gremlin Server and the API for consuming the responses.
+
+The entities are as follows, from the highest level to the lowest:
+
+#### Client
+
+A `Client` represents the entry point to interaction with a Gremlin-supported server. A URL parameter is required for construction, with additional configuration options such as HTTP headers and TLS configuration available.
+
+The `Client` has two main responsibilities:
+
+* Handles initialization and configuration.
+* Handles creation and allocation (i.e. pooling) of `connection` types.
+
+##### Cardinalities
+
+* One `connection` (as of milestone 3).
+  * Will own multiple connections with eventual connection pooling logic.
+
+##### Lifecycle and States
+
+* The `Client` does not track or have any real state.
+* However, `Close()` can be invoked on a `Client` in order to close any instances of `connection` that it owns.
+
+```mermaid
+classDiagram
+	class Client
+	Client: connection *connection
+	Client: NewClient(host, configurations) Client
+	Client: Close()
+	Client: Submit(traversal) ResultSet
+	Client: submit(bytecode) ResultSet
+```
+
+```mermaid
+sequenceDiagram
+	autonumber
+	User->>gremlingo: NewClient()
+	gremlingo-->>User: *Client
+	User->>Client: Submit()
+	Client-->>User: ResultSet
+	
+```
+
+#### connection
+
+A `connection` represents an individual communication component with a Gremlin Server. A `connection` has the sole responsibility of being the representation of a communication channel to a Gremlin Server, providing the interface for sending requests to said server, as well as holding response `ResultSet` instances to consume responses asynchronously until they are consumed.
+
+##### Cardinalities
+
+* One `protocol`
+* Multiple temporary instances of `ResultSet`
+
+##### Lifecycle and States
+
+* States
+  * `initialized`
+  * `established`
+  * `closed`
+  * `closedDueToError`
+* When `close()` is invoked, set the state to `closed` and also invoke `close()` on the `protocol`.
+
+```mermaid
+classDiagram
+	class connection
+	connection: protocol *protocol
+	connection: results "map[string]ResultSet"
```

Comment:
Is there a reason `map[string]ResultSet` is in quotations?

Reply:
Mermaid doesnt like go syntax, so when this is unquoted it tries to parse it and you can't escape it.